### PR TITLE
Add ability to trace cable paths through circuits.

### PR DESCRIPTION
**Pull request by smk4664**
### Fixes: #76
These changes allow paths to be traced through circuits.  Any created connections to circuits will need to be removed and added back in to take advantage, since CablePaths are cached.

I followed an example from Netbox and added the import when it was being used.  I can move it to the top if that is preferred.

### CODE REVIEW

**Comment by smk4664** [comment]
Looks like a status change can also update cables.  I need to add test cases for this, so I am going to set to WIP.

File: nautobot/docs/models/dcim/cable.md
Comment (glennmatthews) [inline nautobot/docs/models/dcim/cable.md:36]: Is the path caching still an issue with your latest diffs? If so, any ideas as to what would be needed to remove this limitation?

Comment (smk4664) [reply nautobot/docs/models/dcim/cable.md:36]: Yes, it is still a limitation.  This is due to the the Termination being a PathEndpoint.  If only one termination exists, then the CablePath assumes this is desired and sets the Destination as the CircuitTermination. This can be seen in test test_105_interface_to_circuittermination.  If you add the second termination before the first cable, the Destination will be changed from CircuitTermination to None, which allows the path refresh to pick up on the second termination and complete the path once the second cable is added.

Comment (smk4664) [reply nautobot/docs/models/dcim/cable.md:36]: One solution is to change the status of the Cable connected to the first termination, if you decide to later add a second termination. I will update the docs accordingly.

File: nautobot/dcim/models/cables.py
Comment (jathanism) [inline nautobot/dcim/models/cables.py:410]: Can you please comment on this block on what it is doing and why it was needed to address this issue? It's not immediately apparent. Doesn't need to be super verbose but something for future maintainers would be helpful! :)

File: nautobot/dcim/models/cables.py
Comment (jathanism) [inline nautobot/dcim/models/cables.py:364]: Is there a reason why this import is here? I would imagine to avoid a circular import issue? If so, can you please add a brief comment?

File: nautobot/circuits/tests/test_filters.py
Comment (jathanism) [inline nautobot/circuits/tests/test_filters.py:6]: Please alphabetize!

File: nautobot/docs/models/dcim/cable.md
Comment (jathanism) [inline nautobot/docs/models/dcim/cable.md:36]: Are we absolutely certain we can't detect this event using a signal and cause this to happen automatically? This seems like an undesirable user experience. Failing that, could we implement a new button or explicit UI/API action to "rebuild paths" or something like that?

**Review by jathanism — CHANGES_REQUESTED** [review]
Minor questions and changes.

File: nautobot/dcim/models/cables.py
Comment (smk4664) [reply nautobot/dcim/models/cables.py:364]: I can add a brief comment. I was following another file that used import at the bottom instead of the top. I can change this to the top if you would like.

File: nautobot/docs/models/dcim/cable.md
Comment (smk4664) [reply nautobot/docs/models/dcim/cable.md:36]: We should create a ticket to implement a button or something for rebuild paths.  As far as the event/signal, I can try to explain how this happens.

Comment (smk4664) [reply nautobot/docs/models/dcim/cable.md:36]: Since cable terminations are also path endpoints, the following is valid [Interface] -- Cable -- [Circuit Termination]. So, a cable path is created with Interface as the origin and Circuit Termination as the destination. If you then add [Circuit Termination 2] and [Interface 2] and connect them with a Cable, refresh_paths function will look for any paths that include [Interface 2] and [Circuit Termination 2].  Since the first path was valid and only included [Interface] and [Circuit Termination] it will not be found by the refresh_paths query and called for update.

We might be able to write a new signal that could detect this.

We could change Circuit Terminations to not be a Path Endpoint, but that will break other use cases for the Cable Termination.

**Review by steffann — COMMENTED** [review]
Besides changing the trace, do we also want to use the extended path in the UI? NB used to show both the connected endpoint and the circuit traversed. I was the one who initially broke that, so I feel a bit obliged to restore it if possible :)

**Comment by jathanism** [comment]
> Besides changing the trace, do we also want to use the extended path in the UI? NB used to show both the connected endpoint and the circuit traversed. I was the one who initially broke that, so I feel a bit obliged to restore it if possible :)

Yes, that's what I would like to see us arrive at as well.

**Comment by smk4664** [comment]
@steffann Thanks for your suggestion, this sentence just triggered an idea "Can't we trigger a path refresh when the CircuitTermination is saved?".  I am going to try this out later today or tomorrow.

**Comment by smk4664** [comment]
I have pushed a new commit that adds a signal for CircuitTermination and removed the note about manual refreshing of the Cable.

**Comment by smk4664** [comment]
> > Besides changing the trace, do we also want to use the extended path in the UI? NB used to show both the connected endpoint and the circuit traversed. I was the one who initially broke that, so I feel a bit obliged to restore it if possible :)
> 
> Yes, that's what I would like to see us arrive at as well.

@steffann Do you have a screenshot you could share?

**Comment by steffann** [comment]
> @steffann Do you have a screenshot you could share?

No, sorry, my main customer is still at NB2.9 and I am having trouble migrating their setup. While I was working on this my 3D printer created a mess, and while I was fixing that half my VMware cluster went down, so I have been a bit distracted :/

Bad excuses, I know ;)

It's past midnight by now, so I'll create a test setup tomorrow!

File: nautobot/circuits/signals.py
Comment (jathanism) [inline nautobot/circuits/signals.py:22]: Since you're calling `ContentType.objects.get_for_model(CircuitTermination)` twice here, can you please assign it to a variable and reference that twice instead?

File: nautobot/dcim/signals.py
Comment (jathanism) [inline nautobot/dcim/signals.py:21]: Since this behavior can be confusing, can you please add to the docstring here explaining what the `rebuild` argument does?

**Review by jathanism — CHANGES_REQUESTED** [review]

**Comment by smk4664** [comment]
image
@steffann Is this what you were looking for?  Note, that Provider was the word I used for the Provider...

**Comment by steffann** [comment]
> @steffann Is this what you were looking for? Note, that Provider was the word I used for the Provider...

That is exactly what I had in mind :)  Just wondering: what happens when there are multiple circuits in the path? Will it show the first one, the last one, or all of them (preferred option)?

I keep volunteering to help out, but you are all way too fast ;)

**Comment by smk4664** [comment]
> > @steffann Is this what you were looking for? Note, that Provider was the word I used for the Provider...
> 
> That is exactly what I had in mind :) Just wondering: what happens when there are multiple circuits in the path? Will it show the first one, the last one, or all of them (preferred option)?
> 
> I keep volunteering to help out, but you are all way too fast ;)

@steffann It will show them all.  It does separate them by `via` 
image
Cable Path if you are interested.
image

**Comment by steffann** [comment]
Awesome!

File: nautobot/circuits/signals.py
Comment (jathanism) [inline nautobot/circuits/signals.py:3]: I know this is is a trivial style nit, but can you please ditch this extra blank line here?

**Review by jathanism — APPROVED** [review]
Approving with my one minor style nit.

File: nautobot/circuits/tests/test_filters.py
Comment (glennmatthews) [inline nautobot/circuits/tests/test_filters.py:299]: I'm a little surprised/worried that this works - including `status` should be mandatory for Device creation, I believe.

File: nautobot/dcim/tables/template_code.py
Comment (glennmatthews) [inline nautobot/dcim/tables/template_code.py:19]: The globe icon isn't rendering because Nautobot (like NetBox 2.10) doesn't use FontAwesome (`fa`), but instead uses Material Design Icons (`mdi`).

**Review by glennmatthews — APPROVED** [review]
Almost there - thank you for seeing this through!

File: nautobot/dcim/tables/template_code.py
Comment (smk4664) [reply nautobot/dcim/tables/template_code.py:19]: I don't think the globe is needed, I just was coping some code that I found referenced.  I am not so great at classes, should I just remove this or is there something you would suggest from `mdi`

File: nautobot/circuits/tests/test_filters.py
Comment (jathanism) [reply nautobot/circuits/tests/test_filters.py:299]: Status isn't mandatory. It can be null. It is more correct to include it however, but not necessary unless it is needed for a test case.

Comment (glennmatthews) [reply nautobot/circuits/tests/test_filters.py:299]: Interesting. The UI treats it as mandatory so I thought that was true on the backend as well.

File: nautobot/dcim/tables/template_code.py
Comment (glennmatthews) [reply nautobot/dcim/tables/template_code.py:19]: I'm fine with removing it.

File: nautobot/circuits/tests/test_filters.py
Comment (smk4664) [reply nautobot/circuits/tests/test_filters.py:300]: I included it just to be more correct.

File: nautobot/dcim/tables/template_code.py
Comment (smk4664) [reply nautobot/dcim/tables/template_code.py]: Thanks! I removed this.  We still have the via, so it should match up with the screenshot.

**Pull request merged**
